### PR TITLE
ScatterElements operator code gen.

### PR DESCRIPTION
**Pull request by etiotto**
Implement support for the ONNX `ScatterElement` operator:

- [x] verification (verify diagnostic completeness)
- [x] shape inference (should be trivial, but verify)
- [x] initial codegen support
- [x] codegen for negative indices
- [x] add lit test to check code generation
- [x] enable end-to-end tests (backend tests)

Signed-off-by: Ettore Tiotto <etiotto@ca.ibm.com>

### CODE REVIEW

File: src/Dialect/ONNX/ONNXOps.cpp [hunk 1]
```diff
@@ -4538,34 +4538,36 @@ mlir::Operation::result_range ONNXScanOp::scan_outputs() {
...
-  mlir::Value indices = operandAdaptor.indices();
-  mlir::Value updates = operandAdaptor.updates();
-
-  // Won't be able to do any checking at this stage.
-  if (!hasShapeAndRank(data) || !hasShapeAndRank(indices) ||
-      !hasShapeAndRank(updates))
-    return success();
+  if (llvm::any_of(operandAdaptor.getOperands(),
+          [](const Value &op) { return !hasShapeAndRank(op); }))
+    return success(); // Won't be able to do any checking at this stage.
```
Comment (AlexandreEichenberger) [inline src/Dialect/ONNX/ONNXOps.cpp:4543]: getting fancy

**Review by AlexandreEichenberger — APPROVED** [review]
LGTM, thanks for the updates, it reads much better

File: src/Dialect/ONNX/ONNXOps.cpp (hunk 1 above)
Comment (etiotto) [reply src/Dialect/ONNX/ONNXOps.cpp:4543]: Easier to read.

File: test/mlir/onnx/invalid.mlir [hunk 2]
```diff
@@ -67,6 +67,39 @@ func @test_concat_verifier_3(%arg0 : tensor<5x5x1x32xf32>, %arg1 : tensor<5x5x3x
...
+// -----
+
+func @test_constantofshape_verifier_4() -> tensor<2xi64> {
+   // expected-error @+2 {{'onnx.ConstantOfShape' op All values of the input tensor must be >=0}}
+   %0 = "onnx.Constant"(){ value = dense<[-1, -2]> : tensor<2xi64> } : () -> tensor<2xi64>
+   %1 = "onnx.ConstantOfShape"(%0) : (tensor<2xi64>) -> tensor<2xi64>
+  "std.return"(%1) : (tensor<2xi64>) -> ()
+}
+
+// -----
```
Comment (etiotto) [inline test/mlir/onnx/invalid.mlir:101]: Moved here to keep the tests in alphabetical order.

File: test/mlir/onnx/invalid.mlir [hunk 3]
```diff
@@ -91,42 +124,42 @@ func @test_onehotencoder_verifier_2(%arg0: tensor<2x2x!onnx.String>) -> tensor<*
 
 // -----
 
-func @test_constantofshape_verifier_1(%arg0: tensor<2x2xi64>) -> tensor<2x2xi64> {
-   // expected-error @+1 {{'onnx.ConstantOfShape' op Input tensor must be a 1D tensor}}
-   %1 = "onnx.ConstantOfShape"(%arg0) : (tensor<2x2xi64>) -> tensor<2x2xi64>
-  "std.return"(%1) : (tensor<2x2xi64>) -> ()
+func @test_pow_verifier_1(%arg0: tensor<1x2x3x4xf32>, %arg1: tensor<f32>) -> tensor<*xf32> {
```
Comment (etiotto) [inline test/mlir/onnx/invalid.mlir:127]: Moved here to keep in alphabetical order.

**Pull request merged**